### ISSUE TITLE
[Bug] Mediastream does not record when browser is minimized.

### ISSUE THREAD

**Issue by Tryptophan** [t-mullen/mediastream-to-webm#4]
How to replicate:
```Javascript
const mtw = require('mediastream-to-webm);
navigator.mediaDevices.getUserMedia({ video: true, audio: true}).then(stream => {
  let encodedStream = mtw.EncodedStream(stream);
  encodedStream.on('data', (data) => {
    console.log(data);
  });
});
```

then minimize your browser, you'll see that the `data` event callback stops.

I believe this has to do with the fact that the `MediaRecorder` has a setInterval timer that's throttled. Any ideas on how to fix this?

**Comment by Tryptophan** [comment]
Maybe using a custom timer and calling `requestData` https://developer.mozilla.org/en-US/docs/Web/API/MediaRecorder/requestData.

**Comment by t-mullen** [comment]
We don't use timers anywhere, and neither does the underlying dependency. Maybe Chrome is throttling the `dataavailable` event? I'll look into explicitly requesting the data like that.

Are you playing audio in the minimized tab?

**Comment by Tryptophan** [comment]
I am yes, I have the stream itself added as a `srcObject` to a video element and playing unmuted audio. I've tested it a bit further and you're right, chrome is throttling the `ondataavailable` event. The `requestData` function I mentioned earlier will force the media recorder to publish the `ondataavailable` event as well.

**Comment by Tryptophan** [comment]
@t-mullen I've also found a really easy to use library to use that doesn't require any sort of background audio hack for getting non-throttled timers: https://www.npmjs.com/package/worker-timers. This uses a web worker to do timing with `setTimeout` and `setInterval`.

**Comment by Tryptophan** [comment]
I've been testing some more and found even though you can force the media recorder to emit `ondataavailable`, the data from the `BlobEvent` will be empty when the tab is blurred/window minimized. See the attached screenshot, the `Uint8Array` from the event is empty when I leave the tab:
![image](https://user-images.githubusercontent.com/4715522/60387570-d0b14480-9a6a-11e9-8909-58af94f2d9d7.png)

**Comment by t-mullen** [comment]
It should be empty some of the time if you're forcing the event faster than the recorder has data. Is it always empty if the tab is minimized?

**Comment by t-mullen** [comment]
Using WebWorkers is an interesting idea - I'd have to look into that.

**Comment by t-mullen** [comment]
I can't actually reproduce this on OSX. The `data` event seems to continue as long as I have audio playing, minimized or not.

If you could tell me what version of windows you're on, I'll try it on a VM this week.

**Comment by Tryptophan** [comment]
It turns out that this is only true for canvases, not video elements. It is always length zero when minimized and using a canvas, even when rendering is done using a non `requestAnimationFrame` timer. You should be able to reproduce with a canvas on MacOS, but my initial statement was incorrect, this works fine for just regular media streams from video/audio elements.

**Comment by t-mullen** [comment]
So you're calling `canvas.captureStream()` and then trying to encode/decode that mediastream?

**Comment by Tryptophan** [comment]
Yes, I'm also rendering in the background using that worker-timer package I sent earlier, I know I'm rendering the canvas as well because there are logging statements that still happen when the tab is blurred, but I think the media stream itself is empty for some external reason.

**Comment by Tryptophan** [comment]
@t-mullen Have you heard anything about this? I think the only thing that can fix this is an update on chrome/Firefox to allow rendering of canvas elements in blurred tabs, which I'm pretty sure is not in the roadmap for either unfortunately.

**Comment by t-mullen** [comment]
Actually - another change to the MediaStreamRecorder API means this module is completely broken until we can find a way to stream the new block format (if that's possible).

**Comment by soundquiet** [comment]
> Yes, I'm also rendering in the background using that worker-timer package I sent earlier, I know I'm rendering the canvas as well because there are logging statements that still happen when the tab is blurred, but I think the media stream itself is empty for some external reason.

@Tryptophan is your problem solved? now i am in the same trouble...

**Comment by t-mullen** [comment]
@soundquiet This module doesn't currently work due to a change in the MediaStreamRecorder API.

**Comment by Tryptophan** [comment]
@t-mullen I found this thread when looking into the webm encoding problem: https://github.com/guest271314/MediaFragmentRecorder/issues/8 Looks like ts-ebml may be able to handle the encoding/decoding of fragmented MediaRecorder chunks into something playable.

**Comment by t-mullen** [comment]
Works fine in latest Chrome again, seems they reverted back to SimpleBlocks.

Still working on streaming the new block format used by Firefox.

**Comment by t-mullen** [comment]
Firefox has actually reverted to SimpleBlock format as well, it just seems unable to play it's own MediaRecorder streams. Tracking here: #7 since this issue is about minimization.

**Comment by Tryptophan** [comment]
@t-mullen Could you iterate on that a bit more? When you say Firefox can't play its own recording is that specific to the same tab or the same browser instance?

**Comment by t-mullen** [comment]
The Webm streams created by Firefox's MediaRecorder can't be played in Firefox's MediaSource. This has always been an issue - this module never worked in Firefox.

I don't know why, but there's an open issue in Bugzilla for it (can't find the link right now).
